lib/hanami/cli.rb

```diff
@@ -45,27 +45,65 @@ def initialize(registry)
     #
     # @param arguments [Array<string>] the command line arguments (defaults to `ARGV`)
     # @param out [IO] the standard output (defaults to `$stdout`)
+    # @param err [IO, NilClass] the standard output or the standard error or nil (default)
+    # @param kernel [Module, NilClass] optional replacement for Kernel, on which #exit() is called.
     #
     # @since 0.1.0
-    def call(arguments: ARGV, out: $stdout)
+    # rubocop:disable Metrics/MethodLength
+    def call(arguments: ARGV, out: $stdout, err: nil, kernel: nil)
+      @kernel = kernel || Kernel
+
+      original_argv = ARGV.dup
+
       result = commands.get(arguments)
 
       if result.found?
-        command, args = parse(result, out)
-
-        result.before_callbacks.run(command, args)
-        command.call(args)
-        result.after_callbacks.run(command, args)
+        call_command(result, out)
       else
-        usage(result, out)
+        exit_code = 0
+        invalid_commands = original_argv.reject { |a| a.start_with?('-') }
+        if invalid_commands.size.positive?
+          report_unknown_command(err, invalid_commands)
+          exit_code = 1
+        end
+        usage(result, out, exit_code)
       end
     end
+    # rubocop:enable Metrics/MethodLength
 
     private
 
+    # Reports unknown commands to the output
+    #
+    # @param err [IO, NilClass] the standard output or the standard error or nil (default)
+    # @param invalid_commands [Array<String>] array of invalid commands
+    #
     # @since 0.1.0
     # @api private
-    attr_reader :commands
+    def report_unknown_command(err, invalid_commands)
+      err&.puts("Error:\n")
+      err&.puts("  Unknown command(s): #{invalid_commands.join(', ')}")
+      err&.puts
+    end
+
+    # Calls the command and runs callbacks.
+    #
+    # @param result [Hanami::CLI::CommandRegistry::LookupResult]
+    # @param out [IO] sta output
+    #
+    # @since 0.1.0
```

Comment:
```suggestion
    # @since x.x.x
```
The version will be filled in automatically if we do this